File: shared/clients/typescript/messagebus/src/nats-backend.ts
import { jwtAuthenticator, wsconnect, type NatsConnection, Msg as NatsMsg } from '@nats-io/transport-node';
import { MessageBus, Subscription, Handler, SubscribeOptions, NATSConfig } from './types';
import { MessageImpl } from './msg';

/**
 * NATS backend implementation
 */
export class NATSBackend implements MessageBus {
  private connection: NatsConnection | null = null;
  private logger: Console;

  constructor(
    private config: NATSConfig,
    logger?: Console
  ) {
    this.logger = logger || console;
  }

  async connect(): Promise<void> {
    if (this.connection) {
      return;
    }

    try {
      const options: any = {
        name: this.config.name || 'messagebus-client',
        servers: this.config.url || 'tls://connect.ngs.global'
      };

      // Add JWT authentication if provided
      if (this.config.jwt && this.config.nkeySeed) {
        const authenticator = jwtAuthenticator(
          this.config.jwt,
          Buffer.from(this.config.nkeySeed)
        );
        options.authenticator = authenticator;
      }

      this.connection = await wsconnect(options);
      this.logger.log('Connected to NATS', { url: options.servers, name: options.name });
    } catch (error) {
      this.logger.error('Failed to connect to NATS:', error);
      throw error;
    }
  }

  async publish(subject: string, data: Uint8Array): Promise<void> {
    if (!this.connection) {
      await this.connect();
    }

    if (!this.connection) {
      throw new Error('NATS connection not available');
    }

    this.connection.publish(subject, data);
    this.logger.debug?.('Published message', { subject, size: data.length });
  }

  async subscribe(subject: string, handler: Handler, opts?: SubscribeOptions): Promise<Subscription> {
    if (!this.connection) {
      await this.connect();
    }

    if (!this.connection) {
      throw new Error('NATS connection not available');
    }

    // Wrap the handler to convert NATS messages to our interface
    const natsHandler = (msg: NatsMsg) => {
      const wrappedMsg = new MessageImpl(msg.subject, msg.data);
      handler(wrappedMsg);
    };

    const subscription = this.connection.subscribe(subject, natsHandler);
    this.logger.debug?.('Subscribed to subject', { subject });

    return new NATSSubscription(subscription, this.logger);
  }

  async close(): Promise<void> {
    if (this.connection) {
      await this.connection.close();
      this.connection = null;
      this.logger.log('NATS connection closed');
    }
  }

  asNATS(): NatsConnection | null {
    return this.connection;
  }
}

/**
 * NATS subscription wrapper
 */
class NATSSubscription implements Subscription {
  constructor(
    private natsSubscription: any,
    private logger: Console
  ) {}

  async unsubscribe(): Promise<void> {
    if (this.natsSubscription) {
      this.natsSubscription.unsubscribe();
      this.logger.debug?.('Unsubscribed from subject');
    }
  }

  async drain(): Promise<void> {
    if (this.natsSubscription && this.natsSubscription.drain) {
      await this.natsSubscription.drain();
      this.logger.debug?.('Drained subscription');
    } else {
      // Fallback to unsubscribe if drain not available
      await this.unsubscribe();
    }
  }
}

File: shared/clients/messagebus/nats_backend.go
package messagebus

import (
        "context"
        "fmt"
        "log/slog"

        "github.com/nats-io/nats.go"
        "github.com/nats-io/nkeys"
)

// natsBus implements Bus using NATS
type natsBus struct {
        conn   *nats.Conn
        logger *slog.Logger
}

// natsSubscription wraps a NATS subscription
type natsSubscription struct {
        sub *nats.Subscription
}

func (s *natsSubscription) Unsubscribe() error {
        return s.sub.Unsubscribe()
}

func (s *natsSubscription) Drain() error {
        return s.sub.Drain()
}

func newNATSBus(ctx context.Context, cfg Config) (Bus, error) {
        logger := cfg.Logger
        if logger == nil {
                logger = slog.Default()
        }

        var opts []nats.Option
        opts = append(opts, nats.Name(cfg.NATS.Name))

        // Set up JWT authentication if provided
        if cfg.NATS.JWT != "" && cfg.NATS.NKeySeed != "" {
                kp, err := nkeys.FromSeed([]byte(cfg.NATS.NKeySeed))
                if err != nil {
                        return nil, fmt.Errorf("failed to parse NKey seed: %w", err)
                }

                opts = append(opts, nats.UserJWTAndSeed(cfg.NATS.JWT, cfg.NATS.NKeySeed))
                defer kp.Wipe()
        }

        // Connect to NATS
        url := cfg.NATS.URL
        if url == "" {
                url = nats.DefaultURL
        }

        conn, err := nats.Connect(url, opts...)
        if err != nil {
                return nil, fmt.Errorf("failed to connect to NATS: %w", err)
        }

        logger.Info("connected to NATS", "url", url, "name", cfg.NATS.Name)

        return &natsBus{
                conn:   conn,
                logger: logger,
        }, nil
}

func (n *natsBus) Publish(subject string, data []byte) error {
        return n.conn.Publish(subject, data)
}

func (n *natsBus) Subscribe(subject string, handler Handler, opts ...SubscribeOption) (Subscription, error) {
        // Wrap our Handler to match nats.MsgHandler signature
        natsHandler := nats.MsgHandler(handler)
        
        sub, err := n.conn.Subscribe(subject, natsHandler)
        if err != nil {
                return nil, fmt.Errorf("failed to subscribe to %s: %w", subject, err)
        }

        n.logger.Debug("subscribed to subject", "subject", subject)

        return &natsSubscription{sub: sub}, nil
}

func (n *natsBus) Close() {
        if n.conn != nil {
                n.conn.Close()
                n.logger.Info("NATS connection closed")
        }
}

func (n *natsBus) AsNATS() (*nats.Conn, bool) {
        return n.conn, true
}